### DISCUSSION TITLE
Additional Currency Source

### DISCUSSION THREAD

**Discussion by bpschuck** [finance-quote/finance-quote#239] [Ideas]
I'm currently looking at another free (1000 free requests per month) currency API.
The API source is [CurrencyFreaks](https://currencyfreaks.com/). Their data does include some precious metals and crypto currencies that were lost when AlphaVantage changed their CURRENCY_EXCHANGE_RATE API to require a premium ($$) API key.

I may post to the GnuCash user and devel email lists asking if there would be an interest in utilizing the CurrencyFreak API in addition or in place of the AlphaVantage API.

- Bruce S.

**Comment by rsbrux** [comment]
Here is another possibility:
https://exchangeratesapi.io/
but we already use the Yahoo API for stock quotes.  Wouldn’t it be easier to use that for curency exchange rates as well than to develop yet another API interface?

**Reply by bpschuck** [reply]
I wish their documentation offered the current currencies, crypto, and metals they have available. Plus it appears exchangeratesapi.io has a limit of 250 requests/month. I really don't want to create yet another account just to see what they have available. Ah, their FAQ has a link to https://exchangeratesapi.io/currencies/. I only see Gold (XAU), Silver (XAG), and BitCoin (BTC) available. Better than the current free AV API, but not much.

About F::Q currency rates. While yahoo_json is usual method people usually use for stocks, the default currency exchange rates are from AlphaVantage. This explains many users report having issues once AV made the API used into a premium service. New F::Q release being tentatively pushed to CPAN this weekend fixes that using a different AV API.

Why a different module from another source API? The original, now premium, AV API for exchange rates included crypto and precious metals with the same API. A few users were taking advantage of this. The newer free API is only regular currency to regular currency.

Bruce S.

**Comment by rsbrux** [comment]
Crypto is apparently available from Yahoo, but I don’t know the details.  This blog post lists a dizzying variety of ways to get data from Yahoo: https://algotrading101.com/learn/yahoo-finance-api-guide/
I just thought it might be easier to expand the existing F::Q I/F to Yahoo to fetch currency exchange rates than to develop something for a new and different API.

**Comment by bpschuck** [comment]
Yahoo can do both foreign exchange, crypto and metals (COMEX). 
Examples:

https://finance.yahoo.com/quote/EURCAD=X?p=EURCAD=X&.tsrc=fin-srch

https://finance.yahoo.com/quote/BTC-USD?p=BTC-USD&.tsrc=fin-srch

https://finance.yahoo.com/quote/GC=F?p=GC=F&.tsrc=fin-srch

This would allow that data with no API key needed. Just need to someone to step up and volunteer to write the module. I can do it, but aside from my day job I am in the middle of a few family matters related to my father's passing this past August and his house (now belonging to my sister and I) being damaged by Hurricane Ian last week.

There is a benefit to having multiple APIs, especially when some change from time to time. Users would be able to select a different source to continue getting data while waiting for a fix. Some APIs, like the original AlphaVantage currency exchange API were able to supply different data (physical currency, crypto, and precious metals) using the same API. With the older AV api, I had my own modified GnuCash that treated Crypto the same as physical currency. That broke when AV made that api a premium service.

Bruce S.

**Comment by bpschuck** [comment]
@rsbrux - I have written a currency module using the same URL that YahooJSON utilizes for security quotes. I need to test it a bit more, but it will likely make it into F::Q v1.55.
I am also thinking of setting the preferred currency module based on an environment variable with a default set if the environment variable is unset or null. This allows easy switching without code changes.

**Comment by edward-dauvergne** [comment]
What about the [European Central Bank data](https://www.ecb.europa.eu/stats/policy_and_exchange_rates/euro_reference_exchange_rates/html/index.en.html)?  This is currently limited to 30 currencies, so not everyone can use it.  But they have a simple XML file updated daily:

- https://www.ecb.europa.eu/stats/eurofxref/eurofxref-daily.xml

A little mathematics is required to convert between two non-Euro currencies.  Although limited, this might end up being quite a reliable source.

**Comment by bpschuck** [comment]
> I'm currently looking at another free (1000 free requests per month) currency API. The API source is [CurrencyFreaks](https://currencyfreaks.com/). Their data does include some precious metals and crypto currencies that were lost when AlphaVantage changed their CURRENCY_EXCHANGE_RATE API to require a premium ($$) API key.

Hello @ka-patel,

Would you be interested in writing a `CurrencyRates/CurrencyFreaks.pm` currency module from scratch? 

Bruce S.

**Comment by ka-patel** [comment]
Hi @bpschuck - I'll give it a shot.

**Reply by ka-patel** [reply]
All Done.

**Comment by bpschuck** [comment]
I stumbled upon X-Rates (ex: https://www.x-rates.com/calculator/?from=USD&to=EUR&amount=1) for currency conversions. Not an API, **but** no API key needed!

**Reply by ka-patel** [reply]
Looking to do another module?

**Comment by Adriano-Baldi** [comment]
If anyone is interested, I would like to point out this Bank of Italy website where they can download the daily exchange rates as well as the monthly averages and annual averages of all currencies. 
This is the link in English to the site: https://www.bancaditalia.it/compiti/operazioni-cambi/index.html?com.dotmarketing.htmlpage.language=1
This is the link to the pdf with instructions for data extraction: https://www.bancaditalia.it/compiti/operazioni-cambi/Operating_Instructions.pdf?language_id=1
It is possible to make queries about historical data from 1918 onwards.
It provides the exchange rates, against the euro and the US dollar, on the latest day for which there are
quotations available for all the currencies quoted in the database.

I've been using it for years and it works great.

This is an example:
 Data: 2024-05-21  Currency: AED    Value:                        3.9898000000 
 Data: 2024-05-21  Currency: AFN    Value:                       78.1692000000 
 Data: 2024-05-21  Currency: ALL    Value:                      100.4100000000 
 Data: 2024-05-21  Currency: AMD    Value:                      421.2000000000 
 Data: 2024-05-21  Currency: ANG    Value:                        1.9447000000 
 Data: 2024-05-21  Currency: AOA    Value:                      926.6920000000 
 Data: 2024-05-21  Currency: ARS    Value:                      965.5855000000 
 Data: 2024-05-21  Currency: AUD    Value:                        1.6278000000 
 Data: 2024-05-21  Currency: AWG    Value:                        1.9447000000 
 Data: 2024-05-21  Currency: AZN    Value:                        1.8469000000 
 Data: 2024-05-21  Currency: BAM    Value:                        1.9558300000 
 Data: 2024-05-21  Currency: BBD    Value:                        2.1728000000 
 Data: 2024-05-21  Currency: BDT    Value:                      127.7878000000 
 Data: 2024-05-21  Currency: BGN    Value:                        1.9558000000 
 Data: 2024-05-21  Currency: BHD    Value:                        0.4080000000 
 Data: 2024-05-21  Currency: BIF    Value:                     3126.5930000000 
 Data: 2024-05-21  Currency: BMD    Value:                        1.0864000000 
 Data: 2024-05-21  Currency: BND    Value:                        1.4632000000 
 Data: 2024-05-21  Currency: BOB    Value:                        7.5070000000 
 Data: 2024-05-21  Currency: BRL    Value:                        5.5371000000 
 Data: 2024-05-21  Currency: BSD    Value:                        1.0864000000 
 Data: 2024-05-21  Currency: BTN    Value:                       90.5115000000 
 Data: 2024-05-21  Currency: BWP    Value:                       14.6811000000 
 Data: 2024-05-21  Currency: BZD    Value:                        2.1728000000 
 Data: 2024-05-21  Currency: CAD    Value:                        1.4797000000 
 Data: 2024-05-21  Currency: CDF    Value:                     3039.4657000000 
 Data: 2024-05-21  Currency: CHF    Value:                        0.9884000000 
 Data: 2024-05-21  Currency: CLP    Value:                      964.5100000000 
 Data: 2024-05-21  Currency: CNY    Value:                        7.8610000000 
 Data: 2024-05-21  Currency: COP    Value:                     4148.6100000000 
 Data: 2024-05-21  Currency: CRC    Value:                      556.2911000000 
 Data: 2024-05-21  Currency: CUP    Value:                       26.0736000000 
 Data: 2024-05-21  Currency: CVE    Value:                      110.2650000000 
 Data: 2024-05-21  Currency: CZK    Value:                       24.6820000000 
 Data: 2024-05-21  Currency: DJF    Value:                      193.0760000000 
 Data: 2024-05-21  Currency: DKK    Value:                        7.4617000000 
 Data: 2024-05-21  Currency: DOP    Value:                       63.6953000000 
 Data: 2024-05-21  Currency: DZD    Value:                      146.0253000000 
 Data: 2024-05-21  Currency: EGP    Value:                       50.6687000000 
 Data: 2024-05-21  Currency: ERN    Value:                       16.7088000000 
 Data: 2024-05-21  Currency: ETB    Value:                       62.4087000000 
 Data: 2024-05-21  Currency: FJD    Value:                        2.4546000000 
 Data: 2024-05-21  Currency: FKP    Value:                        0.8544000000 
 Data: 2024-05-21  Currency: GBP    Value:                        0.8544000000 
 Data: 2024-05-21  Currency: GEL    Value:                        2.9663000000 
 Data: 2024-05-21  Currency: GHS    Value:                       15.6479000000 
 Data: 2024-05-21  Currency: GIP    Value:                        0.8544000000 
 Data: 2024-05-21  Currency: GMD    Value:                       73.0800000000 
 Data: 2024-05-21  Currency: GNF    Value:                     9244.8184000000 
 Data: 2024-05-21  Currency: GTQ    Value:                        8.4332000000 
 Data: 2024-05-21  Currency: GYD    Value:                      227.2966000000 
 Data: 2024-05-21  Currency: HKD    Value:                        8.4778000000 
 Data: 2024-05-21  Currency: HNL    Value:                       26.8672000000 
 Data: 2024-05-21  Currency: HTG    Value:                      144.2797000000 
 Data: 2024-05-21  Currency: HUF    Value:                      385.4300000000 
 Data: 2024-05-21  Currency: IDR    Value:                    17383.9800000000 
 Data: 2024-05-21  Currency: ILS    Value:                        3.9892000000 
 Data: 2024-05-21  Currency: INR    Value:                       90.5115000000 
 Data: 2024-05-21  Currency: IQD    Value:                     1423.1840000000 
 Data: 2024-05-21  Currency: IRR    Value:                   452508.0000000000 
 Data: 2024-05-21  Currency: ISK    Value:                      150.1000000000 
 Data: 2024-05-21  Currency: JMD    Value:                      169.5400000000 
 Data: 2024-05-21  Currency: JOD    Value:                        0.7703000000 
 Data: 2024-05-21  Currency: JPY    Value:                      169.8600000000 
 Data: 2024-05-21  Currency: KES    Value:                      142.8638000000 
 Data: 2024-05-21  Currency: KGS    Value:                       95.7662000000 
 Data: 2024-05-21  Currency: KHR    Value:                     4429.3700000000 
 Data: 2024-05-21  Currency: KMF    Value:                      491.9677500000 
 Data: 2024-05-21  Currency: KPW    Value:                        2.3901000000 
 Data: 2024-05-21  Currency: KRW    Value:                     1480.5600000000 
 Data: 2024-05-21  Currency: KWD    Value:                        0.3334000000 
 Data: 2024-05-21  Currency: KYD    Value:                        0.8908000000 
 Data: 2024-05-21  Currency: KZT    Value:                      480.9300000000 
 Data: 2024-05-21  Currency: LAK    Value:                    21652.0000000000 
 Data: 2024-05-21  Currency: LBP    Value:                    97232.8000000000 
 Data: 2024-05-21  Currency: LKR    Value:                      325.4299000000 
 Data: 2024-05-21  Currency: LRD    Value:                      210.2619000000 
 Data: 2024-05-21  Currency: LSL    Value:                       19.6955000000 
 Data: 2024-05-21  Currency: LYD    Value:                        5.2543000000 
 Data: 2024-05-21  Currency: MAD    Value:                       10.7910000000 
 Data: 2024-05-21  Currency: MDL    Value:                       19.1365000000 
 Data: 2024-05-21  Currency: MGA    Value:                     4797.1600000000 
 Data: 2024-05-21  Currency: MKD    Value:                       61.5793000000 
 Data: 2024-05-21  Currency: MMK    Value:                     2281.4000000000 
 Data: 2024-05-21  Currency: MNT    Value:                     3673.2400000000 
 Data: 2024-05-21  Currency: MOP    Value:                        8.7321000000 
 Data: 2024-05-21  Currency: MRU    Value:                       43.1300000000 
 Data: 2024-05-21  Currency: MUR    Value:                       50.2527000000 
 Data: 2024-05-21  Currency: MVR    Value:                       16.7414000000 
 Data: 2024-05-21  Currency: MWK    Value:                     1880.9025000000 
 Data: 2024-05-21  Currency: MXN    Value:                       17.9800000000 
 Data: 2024-05-21  Currency: MYR    Value:                        5.0979000000 
 Data: 2024-05-21  Currency: MZN    Value:                       69.1700000000 
 Data: 2024-05-21  Currency: NAD    Value:                       19.6955000000 
 Data: 2024-05-21  Currency: NGN    Value:                     1609.9144000000 
 Data: 2024-05-21  Currency: NIO    Value:                       40.0037000000 
 Data: 2024-05-21  Currency: NOK    Value:                       11.5830000000 
 Data: 2024-05-21  Currency: NPR    Value:                      144.8184000000 
 Data: 2024-05-21  Currency: NZD    Value:                        1.7810000000 
 Data: 2024-05-21  Currency: OMR    Value:                        0.4177000000 
 Data: 2024-05-21  Currency: PAB    Value:                        1.0864000000 
 Data: 2024-05-21  Currency: PEN    Value:                        4.0657000000 
 Data: 2024-05-21  Currency: PGK    Value:                        4.1833000000 
 Data: 2024-05-21  Currency: PHP    Value:                       63.2940000000 
 Data: 2024-05-21  Currency: PKR    Value:                      302.5218000000 
 Data: 2024-05-21  Currency: PLN    Value:                        4.2553000000 
 Data: 2024-05-21  Currency: PYG    Value:                     8163.3500000000 
 Data: 2024-05-21  Currency: QAR    Value:                        3.9545000000 
 Data: 2024-05-21  Currency: RON    Value:                        4.9743000000 
 Data: 2024-05-21  Currency: RSD    Value:                      117.1632000000 
 Data: 2024-05-21  Currency: RWF    Value:                     1409.5999000000 
 Data: 2024-05-21  Currency: SAR    Value:                        4.0740000000 
 Data: 2024-05-21  Currency: SBD    Value:                        9.2460000000 
 Data: 2024-05-21  Currency: SCR    Value:                       15.6773000000 
 Data: 2024-05-21  Currency: SDG    Value:                      652.0820000000 
 Data: 2024-05-21  Currency: SEK    Value:                       11.5810000000 
 Data: 2024-05-21  Currency: SGD    Value:                        1.4632000000 
 Data: 2024-05-21  Currency: SHP    Value:                        0.8544000000 
 Data: 2024-05-21  Currency: SLE    Value:                       24.7119000000 
 Data: 2024-05-21  Currency: SOS    Value:                      620.9645000000 
 Data: 2024-05-21  Currency: SRD    Value:                       35.3470000000 
 Data: 2024-05-21  Currency: SSP    Value:                     1709.8769000000 
 Data: 2024-05-21  Currency: STN    Value:                       24.5000000000 
 Data: 2024-05-21  Currency: SVC    Value:                        9.5060000000 
 Data: 2024-05-21  Currency: SYP    Value:                    13647.9000000000 
 Data: 2024-05-21  Currency: SZL    Value:                       19.6955000000 
 Data: 2024-05-21  Currency: THB    Value:                       39.4250000000 
 Data: 2024-05-21  Currency: TJS    Value:                       11.7889000000 
 Data: 2024-05-21  Currency: TMT    Value:                        3.8024000000 
 Data: 2024-05-21  Currency: TND    Value:                        3.3802000000 
 Data: 2024-05-21  Currency: TOP    Value:                        2.5472000000 
 Data: 2024-05-21  Currency: TRY    Value:                       34.9752000000 
 Data: 2024-05-21  Currency: TTD    Value:                        7.3661000000 
 Data: 2024-05-21  Currency: TWD    Value:                       35.0676000000 
 Data: 2024-05-21  Currency: TZS    Value:                     2817.9939000000 
 Data: 2024-05-21  Currency: UAH    Value:                       43.2643000000 
 Data: 2024-05-21  Currency: UGX    Value:                     4130.7300000000 
 Data: 2024-05-21  Currency: USD    Value:                        1.0864000000 
 Data: 2024-05-21  Currency: UYU    Value:                       41.8085000000 
 Data: 2024-05-21  Currency: UZS    Value:                    13786.7500000000 
 Data: 2024-05-21  Currency: VES    Value:                       39.6803000000 
 Data: 2024-05-21  Currency: VND    Value:                    27656.0000000000 
 Data: 2024-05-21  Currency: VUV    Value:                      129.6000000000 
 Data: 2024-05-21  Currency: WST    Value:                        2.9590000000 
 Data: 2024-05-21  Currency: XAF    Value:                      655.9570000000 
 Data: 2024-05-21  Currency: XCD    Value:                        2.9333000000 
 Data: 2024-05-21  Currency: XDR    Value:                        0.8199080000 
 Data: 2024-05-21  Currency: XOF    Value:                      655.9570000000 
 Data: 2024-05-21  Currency: XPF    Value:                      119.3317000000 
 Data: 2024-05-21  Currency: YER    Value:                      271.8200000000 
 Data: 2024-05-21  Currency: ZAR    Value:                       19.6955000000 
 Data: 2024-05-21  Currency: ZMW    Value:                       28.0885000000 
 Data: 2024-05-21  Currency: ZWL    Value:                       14.4598000000 

Best regards

Adriano B.

**Comment by bpschuck** [comment]
@Adriano-Baldi ,

Thanks for the input. However, the F::Q currency rates modules return the exchange rate for any two currencies. The sources used currently provide a quick direct exchange rate. While it would be possible to calculate let's say YEN to CAD by a calculation based on each of those rates to USD, with other options available I don't think anyone is going to volunteer to write a currency module based on this data.

**Comment by Adriano-Baldi** [comment]
@bpschuck ,

I agree, I personally use these data for tax purposes as they are the official ones for the European community.
But since I have never seen this source mentioned, I thought that in case of extreme necessity, if other data sources run out, we can take them from here.

Best regards

Adriano B.

**Comment by ka-patel** [comment]
@bpschuck,

I looked at the API document and actually it does provide a direct conversion between two currencies. According to the GET /dailyRates?referenceDate={}[&baseCurrencyIsoCode={}]&currencyIsoCode={}&lang={} has that ability, just like CurrencyFreaks does. I might be able to write it up quickly if there is any interest to include it.

**Reply by bpschuck** [reply]
@ka-patel 

Note the API doc specification of `currencyIsoCode` (https://www.bancaditalia.it/compiti/operazioni-cambi/Operating_Instructions.pdf?language_id=1 - Page 3 - Input Paramaters):

**currencyIsoCode**
Required:Yes
Repeatable:No
Description: The currency ISO code (case insensitive) against which you want the rates.
It can be "EUR", "USD" or "ITL". If the parameter is not specified, or the
value is not valid, the service will return an error message.

As I stated, one would need to calculate the rate of two currencies when both are not a member of the EUR, USD, ITL set.

**Reply by ka-patel** [reply]
Right, that `GET /dailyRates?referenceDate={}[&baseCurrencyIsoCode={}]&currencyIsoCode={}&lang={}` endpoint is not attractive. Looks like calculation for all currencies are done if retrieved using `GET /latestRates?lang={}` end point. Walking the list to find two currencies is trivial in content returned as JSON. 

Example:

```

### from rec: {
###             country => 'INDIA',
###             currency => 'Indian Rupee',
###             eurRate => '90.2135',
###             isoCode => 'INR',
###             referenceDate => '2024-05-22',
###             uicCode => '031',
###             usdExchangeConvention => 'Foreign currency amount for 1 Dollar',
###             usdExchangeConventionCode => 'C',
###             usdRate => '83.2996'
###           }
###   to rec: {
###             country => 'UNITED STATES',
###             currency => 'U.S. Dollar',
###             eurRate => '1.0830',
###             isoCode => 'USD',
###             referenceDate => '2024-05-22',
###             uicCode => '001',
###             usdExchangeConvention => '-',
###             usdExchangeConventionCode => '-',
###             usdRate => '1'
###           }

### to weight  : '0.0120048595671528'
### from weight: '1'
### amount     : '10000'

### amount/from -> delta/result : 9999,
###                               '120.048595671528'

### to/from -> delta/result : '0.987995140432847',
###                           '120.048595671528'
$VAR1 = '120.048595671528';
```

If there is any interest in including it, then I can do a PR for it.

**Comment by bpschuck** [comment]
On 5/22/24 10:22 AM, ka-patel wrote:

> Right, that |GET 
> /dailyRates?referenceDate={}[&baseCurrencyIsoCode={}]&currencyIsoCode={}&lang={}|
> endpoint is not attractive. Looks like calculation for all currencies
> are done if retrieved using |GET /latestRates?lang={}| end point.
> Walking the list to find two currencies is trivial in content
> returned as JSON.

I believe you are missing my point. To get YEN - INR an extra step is 
need to calculate that exchange rate based on their exchange rate with 
USD. Simple algebra I know. But when other sources such as x-rates.com 
are available, why bother. Besides, I was only noting x-rates.com at 
first just keep the URL handy if needed in the future.

**Reply by ka-patel** [reply]
Understood about the extra step needed, which is as you mentioned simple math. It is something today's smallest compute system can handle without breaking a sweat (a bit?). 

I was seeing it from a different perspective: it is from an arm of a government that is making information available; therefore, it is unlikely to change without much scrutiny or go away as it is highly relied upon it. Fidelity of the data is likely to be highest given that data is from a source that is an authority in that area. And less probability that it will be put behind paywall or rate limit compared to a commercial provider. All these attributes would help the currency portion to be on a better footing in F::Q.

**Comment by bpschuck** [comment]
I just ran across https://fxratesapi.com/. It seems that while their site indicates an api key is required, a simple curl command with no key defined works.

`curl https://api.fxratesapi.com/latest`

or even

`curl https://api.fxratesapi.com/latest?base=EUR'

**Comment by ka-patel** [comment]
Their API set has a dedicated conversion endpoint …

 

***@***.***:~$

***@***.***:~$ curl --request GET --url 'https://api.fxratesapi.com/convert?date=2025-01-06&from=USD&to=EUR&format=JSON&places=4&amount=1' | jq .

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current

                                 Dload  Upload   Total   Spent    Left  Speed

100   182  100   182    0     0    943      0 --:--:-- --:--:-- --:--:--   943

{

  "success": true,

  "query": {

    "from": "USD",

    "to": "EUR",

    "amount": 1

  },

  "info": {

    "rate": 0.9706001761

  },

  "historical": false,

  "date": "2025-01-06T00:07:00.000Z",

  "timestamp": 1736122020,

  "result": 0.9706

}

***@***.***:~$

 

I’ll do more testing to see if there some sort of throttling mechanism or not.

**Comment by ka-patel** [comment]
Oh never mind … their pricing page has throttling details:

 

Free

$0/month

1k requests/month

 

60 requests/hour

 

 

It also mentions conversion endpoint is not available in free subscription but it succeeded.

 

From: Kalpesh Patel ***@***.***> 
Sent: Monday, January 06, 2025 10:01 PM
To: 'finance-quote/finance-quote' ***@***.***>; 'finance-quote/finance-quote' ***@***.***>
Cc: 'Mention' ***@***.***>
Subject: RE: [finance-quote/finance-quote] Additional Currency Source (Discussion #239)

 

Their API set has a dedicated conversion endpoint …

 

***@***.***:~$

***@***.***:~$ curl --request GET --url 'https://api.fxratesapi.com/convert?date=2025-01-06&from=USD&to=EUR&format=JSON&places=4&amount=1' | jq .

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current

                                 Dload  Upload   Total   Spent    Left  Speed

100   182  100   182    0     0    943      0 --:--:-- --:--:-- --:--:--   943

{

  "success": true,

  "query": {

    "from": "USD",

    "to": "EUR",

    "amount": 1

  },

  "info": {

    "rate": 0.9706001761

  },

  "historical": false,

  "date": "2025-01-06T00:07:00.000Z",

  "timestamp": 1736122020,

  "result": 0.9706

}

***@***.***:~$

 

I’ll do more testing to see if there some sort of throttling mechanism or not.

**Comment by ka-patel** [comment]
Wanted to mention that Alpaca's API ([https://docs.alpaca.markets/reference/cryptolatestbars-1])  provides Crypto pricing without needing API key. 

```
***@***.***:~$ curl https://data.alpaca.markets/v1beta3/crypto/us/latest/bars?symbols=BTC%2FUSD | jq .

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   122  100   122    0     0    293      0 --:--:-- --:--:-- --:--:--   293

{
  "bars": {
    "BTC/USD": {
      "c": 95253.461,
      "h": 95253.461,
      "l": 95136,
      "n": 1,
      "o": 95136,
      "t": "2025-01-08T22:43:00Z",
      "v": 0.1,
      "vw": 95136
    }
  }
}

***@***.***:~$

```

**Comment by JDLH** [comment]
Does F::Q have a way of specifying that some currencies are retrieved from one source, and other currencies from another?  I had been thinking of writing a module for the Bank of Canada, since it has rates between CAD and the currencies which interest me. Could this be a part of an overall solution?

**Reply by bpschuck** [reply]
> Does F::Q have a way of specifying that some currencies are retrieved from one source, and other currencies from another?

Negative. 

> Could this be a part of an overall solution?

I don't believe so.

**Reply by bpschuck** [reply]
BTW - The currency exchange rate for a Gnucash user's securities or currencies is not done during price retrieval. It is done in the Gnucash supplied `finance-quote-wrapper`.

**Comment by bpschuck** [comment]
BTW - I have written a currency module using TwelveData's API. Key required, but their docs indicate the basic/free key is allowed 800 queries per day.

**Comment by bpschuck** [comment]
@TiemenSch opened [last week](https://github.com/finance-quote/finance-quote/issues/562#issue-4369227103)

Have you considered adding https://frankfurter.dev/ as a data source?

It does not require an API key and aggregates a whole bunch of providers into a single API.

I was really hoping to find something like this to integrate it with GnuCash such that each transaction can get it's own accurate quote with relative ease.

**Reply by bpschuck** [reply]
@TiemenSch

I've been looking at the Frankfurter API. I may write a currency rate module using the `/v2/rate/<Base>/<Quote>` API in the next 1 or 2 weeks.

**Reply by ka-patel** [reply]
It looks very promising API. I was thinking of writing one up (for myself at least).

Might be worth providing provision to pass the source for quote as a variable since it supports filter by provider in rate call by adding it:

 `/v2/rate/<Base>/<Quote>?providers=XXX`

**Reply by bpschuck** [reply]
> Might be worth providing provision to pass the source for quote as a variable since it supports filter by provider in rate call by adding it:

> `/v2/rate/<Base>/<Quote>?providers=XXX`

I'm a big fan of the KISS principle, and coding for the lowest common denominator. Especially from within GnuCash where it won't be possible to **easily** change the providers. On top of that, few users will likely check to see what rates are available from what providers. Just because something _can_ be done, does not necessarily mean it _should_ be done.

**Reply by bpschuck** [reply]
> I've been looking at the Frankfurter API. I may write a currency rate module using the `/v2/rate/<Base>/<Quote>` API in the next 1 or 2 weeks.

@TiemenSch

I uploaded `BPSCHUCK/Finance-Quote-1.69_01.tar.gz` to CPAN. It contains the Frankfurter currency source module, as well as a couple other changes. It may take some time for it to show up in the CPAN indexes.

Like other currency modules you will need to set `FQ_CURRENCY=Frankfurter` to use it within GnuCash.

**Comment by rob-browncc** [comment]
Late to this long-running thread, but for completeness: **UniRateAPI** was merged as a Finance::Quote currency source via #561 on 2026-05-03 (thanks @bpschuck for the review). So the conversation about free no-API-key currency sources gained another option inside F::Q itself.

Quick fact sheet for anyone reading later:
- Free tier (API key required but signup is free, no card)
- 593 currencies — fiat + select crypto
- Single-call snapshot from any base (`/api/rates?base=EUR`), so any direct cross-pair lookup is one query
- Historical rates + commodities are Pro-tier (won't help the GnuCash history backfill case — Yahoo or ECB remain the right fits there)

*Disclosure: I work on UniRateAPI.*
